### PR TITLE
Archived and deleted

### PR DESCRIPTION
**Pull request by alexander-yevsyukov**

### CODE REVIEW

**Comment by alexander-yevsyukov** [comment]
@armiol, the goal of introducing the `RepositoryView` interface is to be able return view-only instances from a repository like `getArchivedView()` or `getDeletedView()`.

Implementations of the views in `load()` would return mixins that would prohibit mutations. For example, an aggregate returned from a deleted view must not accept commands.

File: server/src/main/java/org/spine3/server/aggregate/AggregateRepository.java [hunk 1]
```diff
@@ -306,16 +304,36 @@ private A loadAndDispatch(I aggregateId, CommandId commandId, Message command, C
         return aggregate;
     }
 
-    /** Posts passed events to {@link EventBus}. */
-    private void postEvents(Iterable<Event> events) {
-        for (Event event : events) {
-            eventBus.post(event);
+    private void logConcurrentModification(I aggregateId, Message command, int newEventCount) {
+        log().warn("Detected the concurrent modification of {} {}" +
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregateRepository.java]: A space symbol is missing after `{} {}`.

File: server/src/main/java/org/spine3/server/aggregate/storage/Predicates.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+            return input == null ||
+                    !(input.getArchived() || input.getDeleted());
+        }
+    };
+
+    private Predicates() {
+        // Prevent instantiation of this utility class.
+    }
+
+    public static Predicate<AggregateStatus> isVisible() {
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/storage/Predicates.java]: As long as this is a part of the public API, it needs some JavaDocs.

File: server/src/main/java/org/spine3/server/entity/RepositoryView.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+package org.spine3.server.entity;
+
+import com.google.common.base.Optional;
+
+import javax.annotation.CheckReturnValue;
+
+/**
+ * A view on a repository.
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/RepositoryView.java]: As a part of the public API, this class needs more details and/or samples.

File: server/src/main/java/org/spine3/server/storage/AbstractStorage.java [hunk 4]
```diff
@@ -95,14 +95,19 @@ protected void checkNotClosed() throws IllegalStateException {
...
+     * @return {@code true} if the storage is open for writing (wasn't closed until now),
+     *         {@code false} otherwise
      * @see #close()
      */
     public boolean isOpen() {
         return this.open;
     }
 
     /**
+     * Verifies if the storage is closed.
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/storage/AbstractStorage.java]: `if` should be `that` or `<empty>`.

File: server/src/main/java/org/spine3/server/storage/AbstractStorage.java [hunk 5]
```diff
@@ -95,14 +95,19 @@ protected void checkNotClosed() throws IllegalStateException {
     }
 
     /**
-     * @return {@code true} if the storage is open for writing (wasn't closed until now), {@code false} otherwise
+     * Verifies if the storage is open.
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/storage/AbstractStorage.java]: `if` should be `that` or `<empty>`.

File: server/src/main/java/org/spine3/server/storage/memory/InMemoryAggregateStorage.java [hunk 6]
```diff
@@ -60,62 +62,54 @@ protected InMemoryAggregateStorage(boolean multitenant) {
...
+        getStorage().putEventCount(id, eventCount);
     }
 
-    /** Used for sorting by timestamp descending (from newer to older). */
-    private static class AggregateStorageRecordReverseComparator implements Comparator<AggregateStorageRecord>,
-                                                                            Serializable {
-        private static final long serialVersionUID = 0L;
+    @Override
+    protected boolean markArchived(I id) {
+        final AggregateStatus currentStatus = getStorage().getStatus(id);
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/storage/memory/InMemoryAggregateStorage.java]: That could lead to the performance issue.

If the aggregate is NOT archived, we'll be calling the storage twice (which is an expensive operation). And even if it is archived, we'll still call the storage at least once.

I would leave a comment somewhere in the JavaDocs that for a real storage (not `InMemory...` one) they should probably always update the state.

File: server/src/main/java/org/spine3/server/storage/memory/InMemoryAggregateStorage.java [hunk 7]
```diff
@@ -60,62 +62,54 @@ protected InMemoryAggregateStorage(boolean multitenant) {
...
+                                                .build());
+        return true;
+    }
 
-            // To define an order:
-            final int firstHashCode = first.hashCode();
-            final int secondHashCode = second.hashCode();
+    @Override
+    protected boolean markDeleted(I id) {
+        final AggregateStatus currentStatus = getStorage().getStatus(id);
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/storage/memory/InMemoryAggregateStorage.java]: Same as above.

File: server/src/main/java/org/spine3/server/storage/memory/TenantRecords.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+import static com.google.common.collect.Maps.newHashMap;
+
+/**
+ * The memory-based storage for {@code EntityStorageRecord} that represents
+ * all storage operations available for data of a single tenant.
+ *
+ * @author Alexander Yevsyukov
+ */
+class TenantRecords<I> implements TenantStorage<I, EntityStorageRecord> {
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/storage/memory/TenantRecords.java]: Do we expect someone using this class from the custom implementations?
It looks like a public API to me.

File: server/src/main/proto/spine/server/aggregate/aggregate_storage.proto [hunk 9]
```diff
@@ -103,3 +103,14 @@ message AggregateEvents {
     // Otherwise, events are full history of the aggregate.
     repeated base.Event event = 2;
 }
+
+// The storage status of an aggregate.
+message AggregateStatus {
+
+    // If `true` the aggregate will not be loaded for regular queries.
+    bool archived = 3;
```
Comment (armiol) [inline server/src/main/proto/spine/server/aggregate/aggregate_storage.proto]: Is it numbered starting from `3` on purpose?

File: server/src/main/java/org/spine3/server/storage/Predicates.java [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+import com.google.common.base.Predicate;
+
+import javax.annotation.Nullable;
+
+/**
+ * Collection of predicates for filtering storage records.
+ *
+ * @author Alexander Yevsyukov
+ */
+public class Predicates {
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/storage/Predicates.java]: It's weird we have `Predicates` in both `org.spine3.server.storage` and `org.spine3.server.aggregate.storage`.

File: server/src/main/java/org/spine3/server/storage/memory/TenantRecords.java (hunk 8 above)
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/storage/memory/TenantRecords.java]: I think we need to be conservative here. We have `InMemory` classes only for testing purposes. 

We may open this class should we see a need, but I'd keep things hidden as much as possible (during the period from v1 to v2) to reduce support calls.

File: server/src/main/proto/spine/server/aggregate/aggregate_storage.proto (hunk 9 above)
Comment (alexander-yevsyukov) [inline server/src/main/proto/spine/server/aggregate/aggregate_storage.proto]: No, it's an omission.

File: server/src/main/java/org/spine3/server/storage/Predicates.java (hunk 10 above)
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/storage/Predicates.java]: These classes cover different inputs. We are going to have `isArchived` and `isDeleted` in both of them soon. What would you suggest to make it better?

File: server/src/main/java/org/spine3/server/storage/memory/InMemoryAggregateStorage.java (hunk 6 above)
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/storage/memory/InMemoryAggregateStorage.java]: @armiol, yes I do agree that there may be performance penalty. Here are questions we need to answer:
* What happens if we don't have a record with an ID? Do we throw? Or, do we just log?
* Do we need to handle repeated archive / delete differently from a business point of view?

File: server/src/main/java/org/spine3/server/entity/RepositoryView.java (hunk 3 above)
Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/entity/RepositoryView.java]: Added.

Comment (alexander-yevsyukov) [inline server/src/main/java/org/spine3/server/entity/RepositoryView.java]: Strangely GitHub doesn't display 849c4f4f5afb40686f95bd5c0b8c11cb12b1f88e.

**Comment by alexander-yevsyukov** [comment]
@armiol, PTAL.

File: server/src/main/java/org/spine3/server/aggregate/AggregateRepository.java [hunk 11]
```diff
@@ -246,14 +250,28 @@ public void store(A aggregate) {
...
 
     /**
-     * Loads the aggregate by the passed ID, or creates it if the aggregate was not found.
+     * Loads the aggregate by the passed ID.
+     *
+     * <p>If the aggregate is not available in the repository this method returns
+     * a newly created aggregate.
+     *
+     * <p>If the aggregate is “invisible” to regular queries, {@code Optional.absent()}
+     * is returned
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/aggregate/AggregateRepository.java]: A dot is missing.

File: server/src/main/java/org/spine3/server/entity/Entity.java [hunk 12]
```diff
@@ -74,24 +75,57 @@
...
+     *
+     * <p>Results of this method call are:
+     * <ul>
+     *   <li>The state object is set to the value produced by {@link #getDefaultState()}.
+     *   <li>The version number is set to zero.
+     *   <li>The {@link #whenModified} field is set to the system time of the call.
+     *   <li>The {@link #status} field is set to the default instance.
+     * </ul>
+     *
+     * This method cannot be called from within {@code Entity} constructor because
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/Entity.java]: `<p>` is missing.

File: server/src/main/java/org/spine3/server/entity/Entity.java [hunk 13]
```diff
@@ -279,6 +303,52 @@ public Timestamp whenModified() {
...
+     * Sets {@code archived} status flag to the passed value.
+     */
+    protected void setArchived(boolean archived) {
+        this.status = getStatus().toBuilder()
+                                 .setArchived(archived)
+                                 .build();
+    }
+
+    /**
+     * Tests weather the entity is marked as deleted.
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/Entity.java]: `whether`.

File: server/src/main/java/org/spine3/server/entity/RepositoryView.java [hunk 14]
```diff
@@ -27,6 +27,13 @@
 /**
  * A view on a repository.
  *
+ * <p>A {@link Repository} may have a view that provides a different set of entities.
+ * For example, there can be views that return archived or deleted entities (that are
+ * not visible by default).
+ *
+ * <p>The default behaviour of loading entities in a {@code Repository} is also a view,
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/entity/RepositoryView.java]: `The behaviour ... is a view` seems to complex to read. Something is missing in the sentence.

File: server/src/main/java/org/spine3/server/storage/AbstractStorage.java [hunk 15]
```diff
@@ -95,7 +95,7 @@ protected void checkNotClosed() throws IllegalStateException {
     }
 
     /**
-     * Verifies if the storage is open.
+     * Tests weather the storage is open.
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/storage/AbstractStorage.java]: `whether`.

File: server/src/main/java/org/spine3/server/storage/AbstractStorage.java [hunk 16]
```diff
@@ -106,7 +106,7 @@ public boolean isOpen() {
     }
 
     /**
-     * Verifies if the storage is closed.
+     * Tests weather the storage is closed.
```
Comment (armiol) [inline server/src/main/java/org/spine3/server/storage/AbstractStorage.java]: `whether`.

File: values/src/main/java/org/spine3/time/Intervals.java [hunk 17]
```diff
@@ -55,6 +56,20 @@ public static Interval between(Timestamp start, Timestamp end) {
     }
 
     /**
+     * Verifies in the timestamp is withing the interval.
```
Comment (armiol) [inline values/src/main/java/org/spine3/time/Intervals.java]: `Verifies if the timestamp is within the interval`.

**Pull request merged**